impl/client/src/main/javascript/web/pentaho/visual/views/echarts/Funnel.js

```diff
@@ -0,0 +1,199 @@
+          fontFamily: font.substring(font.indexOf(" ") + 1)
+        }
+      };
+      options.legend = legend;
+    },
+
+    _buildColors: function() {
+      return this.model.palette.colors.toArray(function(color) {
+        return color.value;
+      });
+    },
+
+    _buildTooltip: function(tooltipFormatString, font) {
+      var toolTip = {
+        formatter: tooltipFormatString,
+        fontSize: font.substring(0, font.indexOf(" ")),
+        fontFamily: font.substring(font.indexOf(" ") + 1)
+      };
+
+      return toolTip;
+    },
+
+    _buildSeries: function() {
+      var dataTable = this.model.data;
+      var range = dataTable.getColumnRange(dataTable.getNumberOfColumns() - 1);
+
+      return [
+        {
+          type: "funnel",
+          left: "20%",
+          right: "20%",
+          top: "10%",
+          bottom: "10%",
+          width: "auto",
+          height: "auto",
+          min: range.min,
+          max: range.max,
+          minSize: 5,
+          maxSize: "100%",
+          sort: this.model.order === "bySizeDescending" ? "descending" : "ascending",
+          legendHoverLink: true,
+          gap: 4,
+          labelLine: {
+            length: 10,
+            lineStyle: {
+              width: 1,
+              type: "solid"
+            }
+          },
+          itemStyle: {
+            borderColor: "#fff",
+            borderWidth: 1
+          },
+          emphasis: {
+            label: {
+              fontSize: 20
+            }
+          }
+        }
+      ];
+    },
+
+    _buildData: function() {
+      var dataTable = this.model.data;
+
+      if(dataTable.originalCrossTable) {
+        dataTable = dataTable.originalCrossTable.toPlainTable({skipRowsWithAllNullMeasures: true});
+      }
+
+      var rowLength = dataTable.getNumberOfRows();
+      var colLength = dataTable.getNumberOfColumns();
+      var font = util.getDefaultFont(null, 12);
+
+      var records = [];
+
+      // Read data from dataTable and push it to records
+      for(var i = 0; i < rowLength; i++) {
+        var tooltipFormatString = "";
+        for(var j = 0; j < colLength; j++) {
+          tooltipFormatString = tooltipFormatString + dataTable.getColumnLabel(j) + " : " +
+            dataTable.getFormattedValue(i, j) + "<br />";
+        }
+        records.push({
+          name: dataTable.getFormattedValue(i, colLength - 2),
+          value: dataTable.getValue(i, colLength - 1),
+          tooltip: this._buildTooltip(tooltipFormatString, font)
+        });
+      }
+
+      return records;
+    },
+
+    _configureData: function(options, records) {
+      options.series.forEach(function(row) {
+        row.data = records;
+      });
+    },
+
+    /** This method initializes the echart library,
+     *configures all the properties, data that are required for the chart and renders it.
+     *
+     * @override */
+    _updateAll: function() {
+      this.base();
```

Comment:
Please add new line after -- two quite different things should be separated by new lines, whenever possible, for readability.